File: packages/node/src/error-kind.ts
import { ERROR_KINDS, type ErrorKind } from './events.js';

/** True when `value` is one of `ERROR_KINDS`. */
export function isErrorKind(value: unknown): value is ErrorKind {
  return typeof value === 'string' && (ERROR_KINDS as readonly string[]).includes(value);
}

/**
 * Best-effort heuristic over free-text error messages. Not a structured
 * code — `success` is the ground truth for pass/fail. See schema/events.md
 * for the documented false-positive mode.
 */
export function classifyError(message: string | undefined | null): ErrorKind | null {
  if (!message) return null;

  if (/not found|does not exist|no such/i.test(message)) return 'not_found';
  if (/\bempty\b|no results?|nothing found|zero results/i.test(message)) return 'empty';
  if (/invalid|required|expected|must be|validation|schema/i.test(message)) return 'validation';
  return 'internal';
}

File: packages/python/src/mcpsignals/error_kind.py
"""Best-effort classification of a free-text error message into a coarse
bucket. This is NOT ground truth - `success` is. See schema/events.md for
the documented false-positive mode.
"""

import re
from typing import Any

from mcpsignals.events import ERROR_KINDS, ErrorKind

_NOT_FOUND = re.compile(r"not found|does not exist|no such", re.IGNORECASE)
_EMPTY = re.compile(r"\bempty\b|no results?|nothing found|zero results", re.IGNORECASE)
_VALIDATION = re.compile(r"invalid|required|expected|must be|validation|schema", re.IGNORECASE)


def is_error_kind(value: Any) -> bool:
    """True when `value` is one of `ERROR_KINDS`."""
    return isinstance(value, str) and value in ERROR_KINDS


def classify_error(message: str | None) -> ErrorKind | None:
    """Bucket an error message into not_found / empty / validation / internal.

    Returns None when there is no message to classify (i.e. the call succeeded).
    """
    if not message:
        return None
    if _NOT_FOUND.search(message):
        return "not_found"
    if _EMPTY.search(message):
        return "empty"
    if _VALIDATION.search(message):
        return "validation"
    return "internal"
